### DISCUSSION TITLE
Possible Inconsistency in `bulk_insert_mappings` with `return_defaults=True`

### DISCUSSION THREAD

**Discussion by krepysh** [sqlalchemy/sqlalchemy#11660] [Usage Questions]
Hello,

I've encountered a potential inconsistency with the `bulk_insert_mappings` method in SQLAlchemy when using the `return_defaults=True` parameter. We have been using this method to insert multiple mappings into the database and retrieve the IDs simultaneously. Here's an example illustrating this use case, which works well with any version in the 1.4 branch.

```python
# This example works fine with sqlalchemy 1.4
from sqlalchemy import create_engine, Column, Integer, String
from sqlalchemy.orm import declarative_base, Session

Base = declarative_base()

class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)
    data = Column(String(10))

engine = create_engine("sqlite://")
Base.metadata.create_all(engine)

session = Session(engine)

mapping = {'data': 'data'}
session.bulk_insert_mappings(A, [mapping], return_defaults=True)
session.commit()

# prints 1
print(mapping['id'])
```

However, with the same bulk_insert_mappings call in the 2.0 branch, the original mappings remain unmodified.

```python
from __future__ import annotations
from sqlalchemy import create_engine, String
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, Session

class Base(DeclarativeBase):
    pass

class A(Base):
    __tablename__ = "a"
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str] = mapped_column(String(10))

engine = create_engine("sqlite://")
Base.metadata.create_all(engine)

session = Session(engine)

mapping = {'data': 'data'}
session.bulk_insert_mappings(A, [mapping], return_defaults=True)
session.commit()

print(mapping)
# prints {'data': 'data'}, no 'id' key!
```

I am uncertain if this behavior should be raised as an issue. Here are my considerations against raising it:

1. The documentation does not explicitly state that the mappings will be modified, in either the 1.x or 2.x branches.
2. In the 2.x branch, bulk_insert_mappings is marked as deprecated.

Despite this, I personally find it beneficial for this method to function consistently across both versions.

If this is indeed a bug and warrants a fix, I would be happy to contribute to a solution.

Thank you for your time and attention.

Best regards,
Anton Topchii

**Comment by zzzeek** [comment]
so that's correct this use case was never anticipated and has no tests.

this patch fails typing but is essentially what changed 

```diff
diff --git a/lib/sqlalchemy/orm/bulk_persistence.py b/lib/sqlalchemy/orm/bulk_persistence.py
index 2ed6a4bea..7d44a3ed7 100644
--- a/lib/sqlalchemy/orm/bulk_persistence.py
+++ b/lib/sqlalchemy/orm/bulk_persistence.py
@@ -125,7 +125,7 @@ def _bulk_insert(
         else:
             mappings = [state.dict for state in mappings]
     else:
-        mappings = [dict(m) for m in mappings]
+        mappings = list(mappings) #[dict(m) for m in mappings]
         _expand_composites(mapper, mappings)
 
     connection = session_transaction.connection(base_mapper)

```

if you're on 2.0 though any reason you are hesitating to move to the more modern [explicit insert version](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#getting-new-objects-with-returning) ?

**Reply by CaselIT** [reply]
it was explicitly changed in this commit https://github.com/sqlalchemy/sqlalchemy/commit/a8029f5a7e3e376ec57f1614ab0294b717d53c05#diff-9d31a091e899d8ccdb290566b4125ad1c6f1a4c9928173d02120b93a8d628451

**Reply by krepysh** [reply]
Most likely we will adopt new insert eventually. Just wanted to share this finding, as well.

**Reply by zzzeek** [reply]
the typing for the argument shows that the incoming list is of mutable dictionaries so it can be supported here, I agree it seems pointless to accept return_defaults with dictionaries otherwise so this is likely a bug

**Comment by CaselIT** [comment]
> In the 2.x branch, bulk_insert_mappings is marked as deprecated.

Please note that legacy is not the same as deprecated. Legacy should be intended as "maintenance only"

**Comment by zzzeek** [comment]
#11661
